fw_info structure change

oyvindronningstad opened the pull request:
Bundle a few changes which modify the fw_info structure. These should ideally be bundled so the number of competing formats are as few as possible.

- Allow explicit invalidation of fw_info structs.
- Refactor EXT_APIs to provide more transparent versioning allowing external checking of version requirements. In the process, the fw_info.h file was reordered quite extensively.
- Add more supported offsets for fw_info structs (0x0 and 0x1000).

Additional changes:
- Add tests for fw_info (not complete coverage)

sigvartmh commented on include/fw_info.h:
```diff
@@ -212,12 +226,25 @@ static inline const struct fw_info *fw_info_check(u32_t fw_info_addr)
+#define FW_INFO_OFFSET2 0x400
+#define FW_INFO_OFFSET3 0x800
+#define FW_INFO_OFFSET4 0x1000
+#define FW_INFO_OFFSET_COUNT 5
```
why do we now have 5?

oyvindronningstad replied:
It's one of the changes (Add more supported offsets).

sigvartmh replied:
okei thanks for clearing it up :)

ru-fu commented on include/fw_info.rst:
```diff
@@ -13,10 +13,10 @@ Purpose
 
 The firmware information structure has a 12-byte magic header and a verified binary layout to ensure that the format is portable and identifiable.
-It must be located at one of three offsets from the start of the image: 0x200, 0x400, or 0x800.
+It must be located at one of three offsets from the start of the image: 0x0, 0x200, 0x400, 0x800, 0x1000.
```
one of the following

oyvindronningstad replied:
Thanks, fixed.

oyvindronningstad commented on include/fw_info.rst:
```diff
@@ -29,6 +29,9 @@ It includes the following information:
 * The address through which to boot into the firmware (the vector table address).
   This address is not necessarily the start of the image.
+* A value that must be 0xFFFFFFFF.
+  If this is set to 0 (e.g. by the bootloader), the image can quickly be established as invalid.
```
In practice it needs only be different from `0xFFFFFFFF`.

hakonfam replied:
If the code enforce that non `0xFFFFFFFF` implies invalid, then IMO we should state that in the docs, otherwise it could be considered a bug that `0xFFFFFFF0` is considered invalid.

oyvindronningstad replied:
Updated docs

oyvindronningstad commented on include/fw_info.rst:
```diff
@@ -29,6 +29,9 @@ It includes the following information:
 * The single, monotonically increasing version number of the image.
 * The address through which to boot into the firmware (the vector table address).
   This address is not necessarily the start of the image.
+* A value that must be 0xFFFFFFFF.
```
Can you elaborate? If this check is spoofed during validation, the signature check will still fail.

hakonfam replied:
Since 0xFFFFFFFF is the  value of erased memory, my gut feeling is that it would be easier to spoof this value somehow, compared to a random value 0x2B68C12A

oyvindronningstad replied:
Ok, I think I understand. For it to work in the OTP region, it would have to be something like 

    #define VALID_VAL 0x2B68FFFF
    #define INVALID_VAL 0x2B680000

oyvindronningstad replied:
Implemented now. I also put the valid value in Kconfig.

ru-fu replied:
Does this need a doc update?

oyvindronningstad replied:
Yes, thanks. Updated docs now.

oyvindronningstad commented on subsys/bootloader/bl_validation/bl_validation.c:
```diff
@@ -161,7 +161,7 @@ static bool validate_firmware(u32_t fw_dst_address, u32_t fw_src_address,
 					fwinfo->firmware_size);
 
-		if (retval != -ESIGINV) {
+		if (retval != -EHASHINV) {
```
Yep, you're right :)

oyvindronningstad commented on subsys/fw_info/Kconfig:
```diff
@@ -83,7 +83,7 @@ config FW_INFO_HARDWARE_ID
 	int "Version number of the fw_*_info structs (8 bits)."
 	range 0 255
-	default 1
+	default 2
```
will squash

oyvindronningstad commented:
Created 3 other PRs:

#1386, #1387, #1388

ru-fu commented on include/fw_info.rst:
```diff
@@ -29,6 +29,9 @@ It includes the following information:
   This address is not necessarily the start of the image.
+* A value that must be 0xFFFFFFFF.
+  If this is set to 0 (e.g. by the bootloader), the image can quickly be established as invalid.
+  The bootloader will set this to 0 when the image fails validation in some way, so a costly validation doesn't need to be performed on every boot.
```
```suggestion
  The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.
```

hakonfam commented:
Typo in one of the commits: "Exlicitly"

hakonfam commented on subsys/bootloader/bl_validation/bl_validation.c:
```diff
@@ -113,6 +113,12 @@ static bool validate_firmware(u32_t fw_dst_address, u32_t fw_src_address,
 		return false;
 	}
 
+	if (fwinfo->valid != 0xFFFFFFFF) {
```
Lets use some define instead of a magic value for `0xFFFFFFFF`

oyvindronningstad replied:
Done

hakonfam commented on subsys/fw_info/Kconfig:
```diff
@@ -82,7 +82,7 @@ config FW_INFO_HARDWARE_ID
 	int "Version number of the fw_*_info structs (8 bits)."
 	range 0 255
-	default 1
+	default 2
```
On a side note, should this field even be visible for the users? IE should it have a prompt? IMO this could be considered a hidden option.

oyvindronningstad replied:
Done

hakonfam commented on include/fw_info.h:
```diff
@@ -64,10 +72,10 @@ struct __packed fw_info {
 
 	/* This firmware's ABI getter. */
-	const fw_info_abi_getter abi_out;
+	fw_info_abi_getter abi_out;
```
Why is this no longer `const` ?

oyvindronningstad replied:
Because none of the other struct members are const. All relevant struct instantiations are marked const anyways, so this isn't strictly needed.

oyvindronningstad commented:
> Typo in one of the commits: "Exlicitly"

Done

oyvindronningstad commented:
I refactored the ABI functionality. See the ABI commit.

oyvindronningstad commented:
Also, I separated the fix-commit into a new branch. Will open a separate PR.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:150: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#150: FILE: include/fw_info.h:171:
+	__noinit static const type * name; \

- total: 1 errors, 0 warnings, 897 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

hakonfam commented on subsys/fw_info/fw_info.c:
```diff
@@ -80,3 +92,15 @@ const struct fw_info_abi *fw_info_abi_find(u32_t id, u32_t flags,
+
+#define INVALID_VAL 0xFFFF0000
+
+#ifdef CONFIG_NRFX_NVMC
```
Note this comment might be outdated, I see that I forgot to submit it, and it was done before the re-wriet.

oyvindronningstad replied:
You can want to have FW_INFO without the invalidation logic, which is why I didn't make it `depends on`. Note that FW_INFO has `imply NRFX_NVMC`.

hakonfam commented:
https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/238 needs to be merged just before this gets merged.

hakonfam commented on include/fw_info.h:
```diff
@@ -27,22 +27,36 @@ extern "C" {
- * @retval -EFAULT  abi was NULL.
+
+/**
+ * This struct is used to request an ABI. The bootloader will populate the
```
Isn't it possible for any image to provide an ABI? Here it states `The bootloader will populate`

oyvindronningstad replied:
The "bootloader" here is any image that boots another image :).
I could change it to "A bootloader will...". Suggestions?

oyvindronningstad replied:
This is now described as "a previous boot stage".

frkv commented on include/fw_info.h:
```diff
@@ -79,12 +103,22 @@ struct __packed fw_info {
+OFFSET_CHECK(struct fw_info, abi_in, 40);
+OFFSET_CHECK(struct fw_info, reserved00, 44);
+
+OFFSET_CHECK(struct fw_info_abi_request, abi_id, 0);
```
Could these static asserts be closer to the type declaration? It comes out of order with the other type declaration first fw_info and then fw_info_abi_request

oyvindronningstad replied:
Moved stuff around.

frkv commented on include/fw_info.h:
```diff
@@ -98,8 +132,10 @@ OFFSET_CHECK(struct fw_info, firmware_address,
  * ID/version combination.
```
Comment on a description and  name of this type:
It seems we spend most of the documentation on fw_info_abi talking about what this type does *not* include in long  related sentences. I would name this a "header" in the type name or use some other naming to emphasize that this is not the complete structure for the ABI. 

Alternatively the documentation could state something like this
This struct serves as the header for an ABI, containing only the basic information about the ABI without describing the  implementation details dependent on an ID and version combination.

oyvindronningstad replied:
This has been rewritten, take another look if you haven't already.

frkv commented on include/fw_info.h:
```diff
@@ -311,6 +346,16 @@ const struct fw_info_abi *fw_info_abi_get(u32_t id, u32_t index);
+ *
+ * @note This function needs to have CONFIG_NRF_NVMC enabled.
+ *
+ * @param[in]  fw_info  The info structure to modify.
```
Modify is the mechanism, maybe use "invalidate" to be clear about the what the act of modifying this structure specifically is doing?

oyvindronningstad replied:
Done

frkv commented on include/fw_info.h:
```diff
@@ -214,41 +277,48 @@ static inline const struct fw_info *fw_info_check(u32_t fw_info_addr)
+static const u32_t fw_info_allowed_offsets[] = {
+					FW_INFO_OFFSET0, FW_INFO_OFFSET1,
+					FW_INFO_OFFSET2, FW_INFO_OFFSET3,
+					FW_INFO_OFFSET4};
```
If the FW_INFO_OFFSET is always increasing array, what about making use of a FW_INFO_OFFSET_INVALID_LAST at the end? This would simplify the logic for the count further down.

frkv commented on include/fw_info.h:
```diff
@@ -214,41 +277,48 @@ static inline const struct fw_info *fw_info_check(u32_t fw_info_addr)
-#if (FW_INFO_OFFSET_COUNT != 3) || ((CURRENT_OFFSET) != (FW_INFO_OFFSET0) && \
-				(CURRENT_OFFSET) != (FW_INFO_OFFSET1) && \
-				(CURRENT_OFFSET) != (FW_INFO_OFFSET2))
+#if (FW_INFO_OFFSET_COUNT != 5) \
```
See comment on FW_INFO_OFFSET_INVALID_LAST

frkv commented on include/fw_info.h:
```diff
@@ -259,7 +329,7 @@ static inline const struct fw_info *fw_info_find(u32_t firmware_address)
 
 	for (u32_t i = 0; i < FW_INFO_OFFSET_COUNT; i++) {
```
See comment on FW_INFO_OFFSET_INVALID_LAST

frkv commented on include/fw_info.rst:
```diff
@@ -29,6 +29,9 @@ It includes the following information:
   This address is not necessarily the start of the image.
+* A value that can be modified in place to invalidate the firmware. See :option:`CONFIG_FW_INFO_VALID_VAL`.
+  If this is set to any other value than :option:`CONFIG_FW_INFO_VALID_VAL` (e.g. by the bootloader), the image can quickly be established as invalid.
+  The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.
```
Maybe add a point on this value being guarded from denial-of-service invalidation by a third party

oyvindronningstad replied:
Added a sentence

frkv commented on samples/bootloader/prj.conf:
```diff
@@ -16,6 +17,10 @@ CONFIG_NRF_RTC_TIMER=n
 CONFIG_FPROTECT=y
 CONFIG_SECURE_BOOT_DEBUG=y
+CONFIG_ASSERT=y
+CONFIG_BL_ROT_VERIFY_ABI_PROVIDE=y
```
I have not seen the verb "provide" used in this context. Is this Zephyr nomenclature?

What about _PROVIDED, _ENABLED, _SUPPORT, or _SUPPORTED?

oyvindronningstad replied:
How about instead of _PROVIDE and _USE we have _ENABLED and _REQUIRED?

oyvindronningstad replied:
Implemented this now. Take a look to see if it works.

frkv replied:
Yes. Easier to read, for sure... Accepted

frkv commented on include/fw_info.h:
```diff
@@ -78,75 +60,156 @@ struct __packed fw_info {
+ * initialized pointer that a previous boot stage will populate with a pointer
+ * to a compliant ABI in another image. An ABI fulfills a request if the ID
+ * matches, all flags in the request are set in the ABI, and the version falls
+ * between the minimum and maximum (inclusive). If `required` is true the image
```
Comma needed: If required is true (comma) the image will not function

frkv commented on include/fw_info.h:
```diff
@@ -311,6 +375,16 @@ const struct fw_info_abi *fw_info_abi_get(u32_t id, u32_t index);
 					u32_t min_version, u32_t max_version);
 
+
+/**Invalidate this image by setting the @c valid value to 0x0.
```
Possible difficulty of matching "image" to content inside fw_info (the valid flag)
since the param[in] fw_info is talking about invalidating the info structure

Suggestion:

Function that invalidates an image by manipulating fw_info

@details Setting the @c valid value to 0x0 in fw_info, invalidates the image it represents

oyvindronningstad replied:
Changed to

    /**Invalidate an image by manipulating its fw_info.
     * 
     * @details Invalidation happens by setting the @c valid value to 0x0.

frkv commented on include/fw_info.h:
```diff
@@ -78,75 +60,156 @@ struct __packed fw_info {
+
+
+/**
+ * This struct is used to request an ABI. It contains a pointer to a non-
```
I would suggest a @details field instead of all the details in one long text

/**
 * Structure used to request an ABI
*
* @details ....
*
* @note If `required` is true the image will not function if it has no access to the ABI.
*/

I would also recommend moving the information about the pointer to a non-initialized pointer to the structure member **abi** instead, to improve readibility. If this is done, the ingress could talk about generic functionality of this struct

oyvindronningstad replied:
Added `@details`.

frkv commented on include/fw_info.h:
```diff
@@ -78,75 +60,156 @@ struct __packed fw_info {
+/**
+ * This struct is used to request an ABI. It contains a pointer to a non-
+ * initialized pointer that a previous boot stage will populate with a pointer
+ * to a compliant ABI in another image. An ABI fulfills a request if the ID
```
Not easy to follow the difference and connection between the "previous boot stage", "another image" and "the image" in this explaination

oyvindronningstad replied:
Rewrote this section

frkv commented on include/fw_info.h:
```diff
@@ -27,47 +27,29 @@ extern "C" {
+ * basic metadata about the ABI: A unique identifier, version and flags to be
+ * used for compatibility matching. The payload that comes after this header is
+ * completely implementation dependent for each ABI.
+ * Note: How to use the ABI, such as the signatures of all the functions in the
```
Express as actual @note

frkv commented on samples/bootloader/src/main.c:
```diff
@@ -46,14 +46,15 @@ extern u32_t _vector_table_pointer;
+		fw_info->image_address);
 
-	if (!bl_validate_firmware_local(fw_info->firmware_address,
+	if (!bl_validate_firmware_local(fw_info->image_address,
```
bl_validate_firmware_local is not documented (but not part of this PR)

I would suggest indicating that fw_info_invalidate is required to be run when this is false (as this function is also used by prevalidate so FW can't be invalidated directly)

oyvindronningstad replied:
Added dox for `bl_validate_firmware_local()`

frkv replied:
Looks good

frkv commented on tests/subsys/fw_info/src/main.c:
```diff
@@ -0,0 +1,113 @@
+u32_t dummy_align;
+u8_t memeq_buf1[] = {1, 2, 3, 4, 5, 6, 7, 8}; /* Aligned */
+u8_t memeq_buf2[] = {1, 2, 3, 4, 5, 6, 7, 8}; /* Aligned */
+u8_t dummy_misalign;
```
I don't know if this leads to an unaligned type behind it. Maybe try with byte-string buffers that are unaligned

oyvindronningstad replied:
You were right, thanks! Fixed this now and added asserts for the expected alignment.

ru-fu commented on include/fw_info.rst:
```diff
@@ -29,91 +29,142 @@ It includes the following information:
+* A value that can be modified in place to invalidate the firmware. See :option:`CONFIG_FW_INFO_VALID_VAL`.
+  If this is set to any other value than :option:`CONFIG_FW_INFO_VALID_VAL` (e.g. by the bootloader), the image can quickly be established as invalid.
+  The bootloader will set this to 0 when the image fails validation in some way, so that there is no need to perform a costly validation on every boot.
+  If the firmware is write protected before being booted by the bootloader, only the bootloader will be able to invalidate it.
```
```suggestion
  If the firmware is write-protected before being booted by the bootloader, only the bootloader can invalidate it.
```

oyvindronningstad commented:
Placed some changes into separate PRs: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1678, https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1679

oyvindronningstad commented:
DNM until #1685  is merged.

oyvindronningstad commented:
@ru-fu There's been a few more changes. Here is the diff of fw_info.rst since your last comment (Dec 13):

https://github.com/NordicPlayground/fw-nrfconnect-nrf/compare/d1d5e588ac99bf6372212a3a5aba48fe07127485..c1ded6042cf0038b52f290cc32cd76967168373f#diff-a911f1d849bdca671f0c2e653412ce5a

It's mostly an additional paragraph on binary compatibility.

oyvindronningstad commented on include/fw_info.h:
```diff
@@ -274,6 +280,17 @@ const struct fw_info_ext_api *fw_info_ext_api_get(u32_t id, u32_t index);
 
+/**Invalidate an image by manipulating its fw_info.
+ *
+ * @details Invalidation happens by setting the @c valid value to 0x0.
```
It makes "valid" monospace. http://www.doxygen.nl/manual/commands.html#cmdc

oyvindronningstad commented on subsys/fw_info/fw_info.c:
```diff
@@ -81,3 +83,15 @@ const struct fw_info_ext_api *fw_info_ext_api_find(u32_t id, u32_t flags,
+void fw_info_invalidate(const struct fw_info *fw_info)
+{
+	/* Check if value has been written. */
+	if (fw_info->valid == CONFIG_FW_INFO_VALID_VAL) {
```
Discussed offline. Will leave as is, since this check and the check for valid images are consistent.

oyvindronningstad commented on subsys/fw_info/Kconfig:
```diff
@@ -101,4 +102,10 @@ config FW_INFO_MAGIC_LEN
 
+config FW_INFO_VALID_VAL
+	hex
+	default 0x9102FFFF
```
Thanks, fixed.

oyvindronningstad commented on subsys/fw_info/fw_info.c:
```diff
@@ -81,3 +83,15 @@ const struct fw_info_ext_api *fw_info_ext_api_find(u32_t id, u32_t flags,
 }
 
+
+#define INVALID_VAL 0xFFFF0000
```
Done :)

hakonfam commented on include/fw_info.h:
```diff
@@ -87,81 +69,199 @@ struct __packed fw_info {
+	struct fw_info_ext_api min_ext_api;
+
+	/* The maximum version accepted. */
+	u32_t ext_api_max_version;
```
Could you explain why the min version is a struct while the max version is an int?

oyvindronningstad replied:
`min_ext_api` refers to the minimum implementation, wrt both version and flags.

hakonfam replied:
so the "min" doesn't really count for flags, right? IMHO the "min" here can be confusing for the reader, when this really is a set of requirements, not all of which has to be "bigger", if you get what I mean.

oyvindronningstad replied:
I agree, we should find a different name.
`requirements`?
`request`?
`ext_api` and then document how it should be interpreted?

oyvindronningstad replied:
Done (`ext_api`)

oyvindronningstad replied:
Oops. Not done, need to fix some more.

oyvindronningstad replied:
Done (`request`)

hakonfam commented on include/fw_info.h:
```diff
@@ -248,51 +348,19 @@ static inline const struct fw_info *fw_info_find(u32_t firmware_address)
  */
-const struct fw_info_ext_api *fw_info_ext_api_find(u32_t id, u32_t flags,
-					u32_t min_version, u32_t max_version);
+bool fw_info_ext_api_provide(const struct fw_info *fwinfo, bool test_only);
```
I don't like having code for testing only. Could this be handled in a different way? e.g. macro, creating a wrapper or something?

oyvindronningstad replied:
It refers to testing for the existence of the ext_apis. To be clear, the argument is meant to be used in regular code, not in tests. We can probably give it a better name.

hakonfam replied:
I only assumed that it was only used for testing.

To me "test only" heavily implies that this is only used for testing, but perhaps that is just me :) Up to you if you want to rename it.

oyvindronningstad replied:
I'll try to find another name :)

oyvindronningstad replied:
Done (`provide`)

hakonfam commented on include/fw_info.rst:
```diff
@@ -53,20 +54,35 @@ The header consists of the following information:
+The RAM buffer must not be initialized, or otherwise touched during bootup of the image.
+This is to preserve the value written there before boot.
+The EXT_API system allows a bootloader to parse the image's requests, the parse its (the bootloader's) own, or other images' EXT_APIs to find matches, then populate the RAM buffers with pointers to other images' EXT_APIs.
+The bootloader will in this way also be aware of whether all images have access to their dependencies.
```
Check this sentence.

hakonfam commented on subsys/fw_info/fw_info.c:
```diff
@@ -6,84 +6,150 @@
 
-extern const u32_t _image_rom_start;
-extern const u32_t _flash_used;
+extern const u32_t _image_rom_start[];
```
Maybe we should add a comment here stating that these are from the linkerscript? I guess not all readers will be known with this scheme.

oyvindronningstad replied:
Good idea, added.

oyvindronningstad commented:
> I need to see some tests using the EXT_API feature before approving :)

The existing bl_crypto tests call the crypto functions using EXT_APIs

frkv commented on include/fw_info.rst:
```diff
@@ -53,20 +54,35 @@ The header consists of the following information:
-After booting, an image can call :cpp:member:`ext_api_in` to retrieve EXT_APIs from the image or images that booted it without knowing where they are located.
+An EXT_API request structure is a structure consiting of a description of the requested EXT_API in the form of an EXT_API header, plus a few more pieces of data.
+The structure also contains the location of a RAM buffer into which a pointer to a matching EXT_API can be placed before booting.
+The RAM buffer must not be initialized, or otherwise touched during bootup of the image.
```
Maybe add a helpful hint, like __attribute__((noinit))

oyvindronningstad replied:
Added a sentence.

hakonfam commented:
@oyvindronningstad  please rebase

oyvindronningstad commented:
> @oyvindronningstad please rebase

Done

oyvindronningstad commented:
@ru-fu FYI this is now only missing a final doc review.

oyvindronningstad commented:
> Please check the commit and squash if everything looks okay.

Thanks @ru-fu! I squashed, and added a slight change: https://github.com/NordicPlayground/fw-nrfconnect-nrf/compare/a7704efbcc8b2b6ec5ba83660127efdba221854e..c012ec46c3f3b4e66eaec0f26d8a2a283aefea40

ru-fu commented on include/fw_info.rst:
```diff
@@ -82,38 +107,86 @@ Creating EXT_APIs
+
+      #define MY_EXT_API_ID 0xBEEF
+
+#. Create a Kconfig entries using :file:`Kconfig.template.fw_info_ext_api`:
```
```suggestion
#. Create Kconfig entries using :file:`Kconfig.template.fw_info_ext_api`:
```

oyvindronningstad replied:
Done :)

hakonfam commented:
Please rebase, LETS DO THIS 💃

oyvindronningstad commented:
> Please rebase, LETS DO THIS 💃

Still blocked by https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1750 :D

tejlmand commented:
DNM until https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/238 is merged, and `west.yml` has been updated to correct SHA

oyvindronningstad commented:
> DNM until [NordicPlayground/fw-nrfconnect-zephyr#238](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/238) is merged, and `west.yml` has been updated to correct SHA

The plan is to wait for #1750 at which point [NordicPlayground/fw-nrfconnect-zephyr#238] is no longer needed.

oyvindronningstad commented:
Rebased now that #1750 is merged.

lemrey commented:
Update the changelog if needed :)

oyvindronningstad commented:
> Update the changelog if needed :)

Thanks :), done
